promql/engine_test.go

```diff
@@ -542,234 +711,496 @@ load 10s
+			// Testing the inner subquery timestamp (since vector selector does not have @)
+			// Inner sum for subquery [100s:25s] @ 200 are
+			// at 100 = 9+10, at 125 = 12, at 150 = 14+15, at 175 = 17, at 200 = 19+20.
+			query: `sum_over_time(sum_over_time(metric{job="1"}[10s])[100s:25s] @ 200)[3s:1s] @ 50`,
+			start: 100,
+			result: Matrix{
+				Series{
+					Points: []Point{{V: 116, T: 47000}, {V: 116, T: 48000}, {V: 116, T: 49000}, {V: 116, T: 50000}},
+					Metric: labels.FromStrings("job", "1"),
+				},
+			},
+		}, {
+			query: `metric_topk and topk(1, sum_over_time(metric_topk[50s] @ 100))`,
+			start: 50, end: 80, interval: 10,
+			result: Matrix{
+				Series{
+					Points: []Point{{V: 995, T: 50000}, {V: 994, T: 60000}, {V: 993, T: 70000}, {V: 992, T: 80000}},
+					Metric: lblstopk3,
+				},
+			},
+		}, {
+			query: `metric_topk and topk(1, sum_over_time(metric_topk[50s] @ 5000))`,
+			start: 50, end: 80, interval: 10,
+			result: Matrix{
+				Series{
+					Points: []Point{{V: 10, T: 50000}, {V: 12, T: 60000}, {V: 14, T: 70000}, {V: 16, T: 80000}},
+					Metric: lblstopk2,
+				},
+			},
+		}, { // timestamp() takes the time of the sample and not the evaluation time.
+			query: `timestamp(metric{job="2"} @ 50)[100s:25s]`,
+			start: 100,
+			result: Matrix{
+				Series{
+					Points: []Point{{V: 50, T: 0}, {V: 50, T: 25000}, {V: 50, T: 50000}, {V: 50, T: 75000}, {V: 50, T: 100000}},
+					Metric: labels.FromStrings("job", "2"),
+				},
+			},
+		}, { // With nesting of timestamp(), the inner timestamp takes the eval time for it's result.
+			query: `timestamp(timestamp(metric{job="2"} @ 50))`,
+			start: 100, end: 130, interval: 10,
+			result: Matrix{
+				Series{
+					Points: []Point{{V: 100, T: 100000}, {V: 110, T: 110000}, {V: 120, T: 120000}, {V: 130, T: 130000}},
+					Metric: labels.FromStrings("job", "2"),
+				},
+			},
+		}, { // timestamp() takes the time of the sample and not the evaluation time.
+			query: `(time() + timestamp(metric{job="2"} @ 50))[100s:25s]`,
+			start: 100,
+			result: Matrix{
+				Series{
+					Points: []Point{{V: 50, T: 0}, {V: 75, T: 25000}, {V: 100, T: 50000}, {V: 125, T: 75000}, {V: 150, T: 100000}},
+					Metric: labels.FromStrings("job", "2"),
+				},
 			},
-			Start: time.Unix(1, 0),
 		},
+	}
+
+	for _, c := range cases {
+		t.Run(c.query, func(t *testing.T) {
+			if c.interval == 0 {
+				c.interval = 1
+			}
+			start, end, interval := time.Unix(c.start, 0), time.Unix(c.end, 0), time.Duration(c.interval)*time.Second
+			var err error
+			var qry Query
+			if c.end == 0 {
+				qry, err = test.QueryEngine().NewInstantQuery(test.Queryable(), c.query, start)
+			} else {
+				qry, err = test.QueryEngine().NewRangeQuery(test.Queryable(), c.query, start, end, interval)
+			}
+			require.NoError(t, err)
+
+			res := qry.Exec(test.Context())
+			require.NoError(t, res.Err)
+			require.Equal(t, c.result, res.Value, "query %q failed", c.query)
+		})
+	}
+}
+
+func TestMaxQuerySamples(t *testing.T) {
```

Comment:
Git is getting confused by you moving functions around and not providing a readable diff here. Can you try to fix that so I can review this? I can't presently tell if you're breaking/removing existing tests.

Also, why did you remove the result field? That's important to help verify that we're testing what we think we're testing, and seeing the number of output samples.

Reply:
I actually modified the test case format and also removed some which would be taken care by the automation below. Happy to open another PR for that if that helps.

Why is it important to have the result checked in `TestMaxQuerySamples`? I thought the correctness of results of queries is taken care by other tests, and since this test specifically mentions max query samples, I though keeping it to only check the limit is enough?

Reply:
The new version makes sense to me 👍🏽